### ISSUE TITLE
Dropout during prediction phase

### ISSUE THREAD

**Issue by ProfessorHT** [keras-team/keras#18129]
Hello,

I have used "tf.keras.layers.Dropout(0.2)(x, training=True)" to turn on the dropout during the prediction phase when i trained my model. Is there any method to turn it off again during the prediction when loading the h5 saved model?

I'am using TF version: 2.10.

Thank you

**Comment by tilakrayal** [comment]
@ProfessorHT,
A Keras Model is [trainable by default](https://github.com/keras-team/keras/blob/master/keras/engine/base_layer.py#L154) - you have two means of freezing all the weights:

1. model.trainable = False before compiling the model
2. for layer in model.layers: layer.trainable = False - works before & after compiling

We can use `layer.trainable=False` for the same. Layers & models also feature a boolean attribute trainable. Its value can be changed. Setting **layer.trainable** to False moves all the layer's weights from trainable to non-trainable. This is called "freezing" the layer: the state of a frozen layer won't be updated during training (either when training with fit() or when training with any custom loop that relies on trainable_weights to apply gradient updates).

```
layer = keras.layers.Dense(3)
layer.build((None, 4))  # Create the weights
layer.trainable = False  # Freeze the layer

print("weights:", len(layer.weights))
print("trainable_weights:", len(layer.trainable_weights))
print("non_trainable_weights:", len(layer.non_trainable_weights))
```

Thank you!

**Comment by ProfessorHT** [comment]
Thank you for the response. 
After i trained my model and saved as h5 file, I have try to load the model and use the code below before predicting:
for layer in model.layers:
   if layer == "dropout" : 
      layer.trainable = False
As you mentioned above that this method will turn off the layer named dropout. But if I try to make a test on a single image for 100 time approximately at each time i receive different prediction and it's a huge difference so it's like the dropout layer is still activated in the test time.

**Comment by tilakrayal** [comment]
@ProfessorHT,
Dropout is not used after training when making a prediction with the fit network.

The weights of the network will be larger than normal because of dropout. Therefore, before finalizing the network, the weights are first scaled by the chosen dropout rate. The network can then be used as per normal to make predictions. Thank you!

**Comment by ProfessorHT** [comment]
Hello @tilakrayal 
Thank you for your response. 

> Therefore, before finalizing the network, the weights are first scaled by the chosen dropout rate. The network can then be used as per normal to make predictions. 

That's right. But the idea is that it didn't function like this for me. 

========================Code & Results=================================== 

The test code is stated below:
`BATCH_SIZE = 1
IMG_SIZE = (... , ... )

allevaccuracy = []
allevaloss = []

model = tf.keras.models.load_model("Model.h5")

test_data = tf.keras.utils.image_dataset_from_directory(
    'test',
    label_mode='categorical',
    batch_size=BATCH_SIZE,
    image_size=IMG_SIZE,
)  

for layer in model.layers: 
    layer.trainable = False`


When i try this code on the test dataset that contain two image (for test of the code) I get different results at each run as shown below:
=============================================================================
model.predict(test_data)
2/2 [==============================] - 0s 358ms/step
Out[3]: 
array([[1.1455322e-05, 9.9998856e-01],
       [1.0000000e+00, 5.6772804e-09]], dtype=float32)

model.predict(test_data)
2/2 [==============================] - 0s 249ms/step
Out[4]: 
array([[9.9999988e-01, 1.4690319e-07],
       [1.8595392e-03, 9.9814045e-01]], dtype=float32)

model.predict(test_data)
2/2 [==============================] - 0s 247ms/step
Out[5]: 
array([[4.6881149e-03, 9.9531186e-01],
       [1.0000000e+00, 1.4155777e-08]], dtype=float32)

model.predict(test_data)
2/2 [==============================] - 0s 223ms/step
Out[6]: 
array([[1.0000000e+00, 6.6442278e-09],
       [1.3445069e-05, 9.9998653e-01]], dtype=float32)

model.predict(test_data)
2/2 [==============================] - 0s 355ms/step
Out[7]: 
array([[2.1816030e-04, 9.9978191e-01],
       [9.9997663e-01, 2.3322356e-05]], dtype=float32)

model.predict(test_data)
2/2 [==============================] - 1s 458ms/step
Out[8]: 
array([[1.000000e+00, 3.651782e-08],
       [2.165683e-03, 9.978344e-01]], dtype=float32)

model.predict(test_data)
2/2 [==============================] - 0s 253ms/step
Out[9]: 
array([[9.999994e-01, 5.724883e-07],
       [6.958003e-02, 9.304200e-01]], dtype=float32)
=============================================================================

Also as you stated in the first reponse using 

> for layer in model.layers: layer.trainable = False - works before & after compiling
 
there is no trainable parameter as shown below:

Total params: 23,568,898
Trainable params: 0
Non-trainable params: 23,568,898


========================Question=================================== 
1) Now after you have view the codes and results why the result differ each time? 
2) Is there any other method to apply MC Dropout? (Without using  training=True in "tf.keras.layers.Dropout(0.2)(x, training=True)"  at the training time)
3) What can i do to let the results be the same at each time?
